src/openfermion/measurements/_hamiltonian_stabilizers_sets.py

```diff
@@ -0,0 +1,93 @@
+#   limitations under the License.
+
+"""Tools to split the Hamiltonian into subsets based on the support of
+its stabilizers."""
+
+import numpy
+import copy
+from openfermion.ops import QubitOperator
+
+
+def _has_disjoint_indices(op1, op2):
+    """
+    Auxiliary function of get Hamiltonian subsets.
+
+    Check if two QubitOperators have any qubit in common.
+
+    Args:
+        op1(QubitOperator): Single Pauli string.
+        op2(QubitOperator): Single Pauli string.
+
+    Returns:
+        overlap (Boolean): True if the operator and stabilizer overlap.
+    """
+    indices_in_op1 = {index for (index, _) in list(op1.terms.keys())[0]}
+    indices_in_op2 = {index for (index, _) in list(op2.terms.keys())[0]}
+
+    return indices_in_op1.isdisjoint(indices_in_op2)
+
+
+def get_hamiltonian_subsets(hamiltonian, stabilizers):
+    """
+    Create subsets of Hamiltonian.
+
+    This function splits a Hamiltonian into subsets of Pauli strings that do
+    not overlap with stabilizers.
+    Pauli strings might be repeated between subsets if they do not share any
+    qubit with the stabilizer.
+    The Pauli strings that overlap with all stabilizers will be stored with a
+    separate set.
+    It can be used to perform SV and not disregarding all data if only a single
+    stabilizers signals an error.
+
+    Args:
+        hamiltonian(QubitOperator): Hamiltonian to be split.
+        stabilizers(list or QubitOperator): List of stabilizers as
+            QubitOperators.
+    Return:
+        hamiltonian_subsets(list): List of QubitOperator which are a subset of
+            the Hamiltonian.
+            Each QubitOperator does not overlap with the stabilizer at the same
+            position in the list.
+        remaining_paulis(QubitOperator): Operator that contains the Pauli
+            strings that overlap with all the stabilizers.
+            If there are no paulis left an empty QubitOperator is returned.
+    Raises:
+        TypeError: Input hamiltonian must be QubitOperator.
+        TypeError: Input stabilizer_list must be QubitOperator or list.
+    """
+    if not isinstance(hamiltonian, QubitOperator):
+        raise TypeError('Input terms must be QubitOperator.')
+    if not isinstance(stabilizers, (QubitOperator, list,
+                                    tuple, numpy.ndarray)):
+        raise TypeError('Input stabilizers must be QubitOperator or list.')
+
+    hamiltonian_subsets = []
+    remaining_paulis = copy.deepcopy(hamiltonian)
+    for stab in list(stabilizers):
+        # Initialize subset
+        ham_subset = QubitOperator()
+        for pauli in hamiltonian:
+            if _has_disjoint_indices(pauli, stab) is True:
+                ham_subset += pauli
+                # Try if a Pauli string is still in the remaining set
+                # otherwise continue.
+                try:
+                    remaining_paulis.terms.pop(list(pauli.terms.keys())[0])
+                except:
+                    continue
```

Comment:
Why is the try-except necessary? Isn't the Pauli here guaranteed to be inside `remaining_paulis'?